File: Bronze/2016-03/diamond_open2016.java
// Diamond Collector - USACO Bronze US Open 2016 (http://www.usaco.org/index.php?page=viewproblem2&cpid=639)
// This problem was completed on November 21, 2020, in 27 minutes, with all 10/10 test cases passed (second try)

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.Scanner;

public class diamond_open2016 {
    static int numDiamonds;
    static int sizeDifference;
    static int[] diamondSizes;

    public static int diamond() {
        Arrays.sort(diamondSizes);
        int maxDiamonds = 0;

        for (int i=0; i<numDiamonds; i++) {
            int startingDiamond = diamondSizes[i];
            int currDiamonds = 1;

            for (int j=i+1; j<numDiamonds; j++) {
                int endingDiamond = diamondSizes[j];

                if ((endingDiamond - startingDiamond) > sizeDifference) {
                    break;
                } else {
                    currDiamonds++;
                }
            }

            maxDiamonds = Math.max(maxDiamonds, currDiamonds);
        }

        return maxDiamonds;
    }

    public static void main(String[] args) throws IOException {
        // input
        String problemName = "diamond";
        Scanner sc = new Scanner(new File(problemName + ".in"));

        numDiamonds = sc.nextInt();
        sizeDifference = sc.nextInt();
        diamondSizes = new int[numDiamonds];
        for (int i=0; i<numDiamonds; i++) {
            diamondSizes[i] = sc.nextInt();
        }

        // algorithm
        int maxDiamonds = diamond();

        // output
        PrintWriter out = new PrintWriter(new FileWriter(problemName + ".out"));
        out.println(maxDiamonds);
        out.close();
    }
}

File: Silver/2016-03/diamond_open2016.cpp
// Diamond Collector
// USACO Silver 2016 US Open: http://www.usaco.org/index.php?page=viewproblem2&cpid=643
// XJOI Problem 15262: https://xjoi.net/contest/3353/problem/2?locale=en

#include <bits/stdc++.h>

using namespace std;

int n, k;
int diamonds[50001];
int valid[50001]; // for a valid[i], i = left, valid[i] = diamonds in case, i+valid[i]-1 = right

int main() {
    ifstream fin("diamond.in");
    ofstream fout("diamond.out");

    fin >> n >> k;
    for (int i=1; i<=n; i++) fin >> diamonds[i];

    sort(diamonds+1, diamonds+n+1);

    for (int left=1, right=1; left<=n; left++) {
        while (right <= n && diamonds[right] - diamonds[left] <= k) right++;
        valid[left] = right-left;
    }

    int result = 0;

    for (int i=1; i<=n; i++) {
        for (int j=i+valid[i]; j<=n; j++) {
            result = max(result, valid[i]+valid[j]);
        }
    }

    fout << result << endl;
}